src/ImageSharp/Formats/Jpeg/Components/Decoder/SpectralConverter{TPixel}.cs

```diff
@@ -98,52 +123,45 @@ public Buffer2D<TPixel> GetPixelBuffer(CancellationToken cancellationToken)
+        /// <returns>Scaled jpeg image size.</returns>
+        private Size GetResultingImageSize(Size nativeSize, out int blockPixelSize)
         {
-            MemoryAllocator allocator = this.configuration.MemoryAllocator;
-
-            // iteration data
-            int majorBlockWidth = frame.Components.Max((component) => component.SizeInBlocks.Width);
-            int majorVerticalSamplingFactor = frame.Components.Max((component) => component.SamplingFactors.Height);
-
-            const int blockPixelHeight = 8;
-            this.pixelRowsPerStep = majorVerticalSamplingFactor * blockPixelHeight;
-
-            // pixel buffer for resulting image
-            this.pixelBuffer = allocator.Allocate2D<TPixel>(
-                frame.PixelWidth,
-                frame.PixelHeight,
-                this.configuration.PreferContiguousImageBuffers);
-            this.paddedProxyPixelRow = allocator.Allocate<TPixel>(frame.PixelWidth + 3);
-
-            // component processors from spectral to Rgba32
-            const int blockPixelWidth = 8;
-            var postProcessorBufferSize = new Size(majorBlockWidth * blockPixelWidth, this.pixelRowsPerStep);
-            this.componentProcessors = new JpegComponentPostProcessor[frame.Components.Length];
-            for (int i = 0; i < this.componentProcessors.Length; i++)
+            if (this.targetSize == null)
             {
-                this.componentProcessors[i] = new JpegComponentPostProcessor(allocator, frame, jpegData, postProcessorBufferSize, frame.Components[i]);
+                blockPixelSize = 8;
+                return nativeSize;
             }
+            else
+            {
+                const uint jpegBlockPixelSize = 8;
 
-            // single 'stride' rgba32 buffer for conversion between spectral and TPixel
-            this.rgbBuffer = allocator.Allocate<byte>(frame.PixelWidth * 3);
-
-            // color converter from Rgba32 to TPixel
-            this.colorConverter = this.GetColorConverter(frame, jpegData);
-        }
+                Size targetSize = this.targetSize.Value;
+                int outputWidth = nativeSize.Width;
+                int outputHeight = nativeSize.Height;
+                blockPixelSize = 1;
 
-        /// <inheritdoc/>
-        public override void ConvertStrideBaseline()
-        {
-            // Convert next pixel stride using single spectral `stride'
-            // Note that zero passing eliminates the need of virtual call
-            // from JpegComponentPostProcessor
-            this.ConvertStride(spectralStep: 0);
+                for (int i = 1; i < ScalingFactors.Length; i++)
+                {
+                    int scale = ScalingFactors[i];
+                    int scaledw = (int)Numerics.DivideCeil((uint)(nativeSize.Width * scale), jpegBlockPixelSize);
+                    int scaledh = (int)Numerics.DivideCeil((uint)(nativeSize.Height * scale), jpegBlockPixelSize);
```

Comment:
Should there https://github.com/SixLabors/ImageSharp/blob/136cc143bba2fca255328751dd4580e4f8d413c3/src/ImageSharp/Common/Helpers/Numerics.cs#L909 be an overload for `int` which does the `(uint)` and `(int)` cast?
So the code here becomes cleaner?

Reply:
While it would really clean up call site it would give false promise of working with negative numbers. Even though this method is internal and most (if not all) of the values apart from pixels are positive - I think it's better leave it as is.